Changes Cellrate

atlantiscze opened the pull request:
- Cellrate changed to 1/1800 (1/(3600/machinery controller tick, which is 2 seconds)).

**Why?**
Reason for this is that power cell rating now equals it's actual capacity in watthours. In short, a cell with rating of 1000 can provide 1000W of power over the duration of one hour. It is easier to predict how long a cell might last. Old value was semi-randomly picked one and converting it to watthours required you to do quite some math (convert charge to joules, and joules to watthours).
- Removes SMESRATE define. SMESes now use CELLRATE instead. There is absolutely no reason, as the value is only used internally, and has no effect on actual game.
- Tweaks SMES magnetic coil properties a bit. Mostly results in rounded values. SMES coils are a little bit smaller capacity wise (about 56~ -> 50kWh for regular coil), but it's not marginal enough to have a major gameplay effect.
- Expands the original fractal reactor (infinite powersource) idea. It no longer produces a world-wide message when spawned. Also adds APC dummy load and power cable dummy load, which are useful for testing of power flow and stuff. These debug items also display powernet information when examined. This information may be hidden by setting a var, for example when an admin wants to use varedit resprited variant for an event or something.

NOTE: Cell capacities have not yet been changed. I have a piece of debug code in place that will help me trace most power cells, to ensure they are updated properly. This debug code will be removed later.

atlantiscze committed:
Changes Cellrate
- Cellrate changed to 1/1800 (1/(3600/machinery controller tick, which is 2 seconds)). Cellrate is now therefore in watthours.
- Removes SMESRATE define. SMESes now use CELLRATE instead. There is absolutely no reason, as the value is only used internally, and has no effect on actual game.
- Tweaks SMES magnetic coil properties a bit. Mostly results in rounded values. SMES coils are a little bit smaller capacity wise (about 56~ -> 50kWh for regular coil), but it's not marginal enough to have a major gameplay effect.
- Expands the original fractal reactor (infinite powersource) idea. It no longer produces a world-wide message when spawned. Also adds APC dummy load and power cable dummy load, which are useful for testing of power flow and stuff. These debug items also display powernet information when examined. This information may be hidden by setting a var, for example when an admin wants to use varedit resprited variant for an event or something.

NOTE: Cell capacities have not yet been changed. I have a piece of debug code in place that will help me trace most power cells, to ensure they are updated properly. This debug code will be removed later.

Raptor1628 commented:
What will be the most noticeable player-side change from this?

atlantiscze commented:
Two most noticeable changes are:
- A power cell with (for example) 1000 rating will hold more energy (1 800 000J).
- Power cell ratings will be re-balanced (read: reduced) accordingly. The amount of energy power cells will hold will be more or less same as it is now (except for the fact that i will round the numbers). I also want to make higher capacity cells a bit rarer, and the crappy cells (those you can find in APCs for example) much more common.

GinjaNinja32 commented:
example map uses a fractal reactor for power, will need the path changing for Travis to pass.

Hubblenaut commented:
Looking at the code, that is actually not converting to Joules at all. Code shouldn't be measuring with that anyway unless it takes ticklag in account to properly convert it.

PsiOmegaDelta commented:
In what way does ticklag affect the machine/obj process update intervals that power consumption does no longer fairly strictly correlate with them?*

In either case, we have a proc for acquiring the configured schedule interval for a given process, instead of hoping it will remain constant forever and never changed:
 `process_schedule_interval("name_of_process")` 
Yes, it's awful that it takes a string, would much prefer a type, but that's currently how it is.

\* The exception being when the process realize the current tick is about to be over-extended, and sleep for a bit.

atlantiscze commented:
@Hubblenaut 

Joules != Watthours. I have done tests, and cell with 1000 Wh of capacity under the new system indeed lasts for about one hour (plus minus few minutes, taking lag in mind) with load of 1000W.

Our machinery ticks once per two seconds by default, and only inaccuracy problems that may arise would be when server is lagging so heavily, that it can't tick once per two seconds. Either way, other power related code does not use real time (it works on per-tick basis), so why should this do? It provides time in watthours, assuming the server is not overloaded, which is majority of cases.

Hubblenaut commented:
> Joules != Watthours

Joules/Watthous/Wattminutes all describe the same and are dependant on the time that has passed since the last tick fired.

When I changed SMES' to use Kilowatthours as units, all I did was to change the constants to make it resemble kilowatthours as closely as possible, but in no way exact. It's not a nice solution, it just happened to be the easiest at the time.

Should the ticklag constant be changed, the way it's calculated wouldn't make sense anymore. As per definition, energy calculates as by work times time, which in the environment of a game should be the game time, which is the amount of ticks fired mutliplied with the ticklag, which is none other than a conversion from tick to seconds.

(Unless of course I'm wrong here and our game time is not actually calculated with ticklag. In that case, disregard.)

The energy cells right now are, opposed to the SMES', making use of arbitrary energy units. If they happen to have 1000 units of energy lasting for exactly 1 hour with a load of 1000 W, then that's a surprising coincidence.

GinjaNinja32 commented:
Regardless of the current situation, a move towards a more believable and understandable setup isn't a bad thing. SMESes and cells having different units is not a good situation for player understanding.

mwerezak commented on code/__defines/machinery.dm:
```diff
@@ -4,9 +4,7 @@ var/global/defer_powernet_rebuild = 0      // True if net rebuild will be called
-#define CELLRATE 0.002 // Multiplier for watts per tick <> cell storage (e.g., 0.02 means if there is a load of 1000 watts, 20 units will be taken from a cell per second)
-                       // It's a conversion constant. power_used*CELLRATE = charge_provided, or charge_used/CELLRATE = power_provided
-#define SMESRATE 0.05  // Same for SMESes. A different number for some reason.
+#define CELLRATE (1/1800) // Multiplier for charge units. Converts cell charge units(watthours) to joules. Takes into consideration that our machinery ticks once per two seconds.
```
If this depends on the machinery tick rate to be correct then you should create a #define for the machinery tick rate and reference it where it is needed (this has less overhead than embedding a proc call into the define).

atlantiscze committed:
Fixes example map compilation

atlantiscze committed:
Adjusts various power cell ratings
- Removes mapped-in powercells with manually edited maxcharge variables.
- Adds a define for machinery tickrate, so cellrate may be calculated without performance drawbacks that come from usage of a proc in define.
- Goes through various machines, guns and other things that use custom capacity power cells. Reduces the capacities accordingly.
- Tweaks random item spawner probabilities. The intention is to make high capacity power cells much more rare in comparison to current state, where they are almost exclusively used.
- Removes high capacity power cells from various vending machines. Replaced with standard power cells. EngiVend still provides high capacity cells in a limited amount when hacked.
- Vending machines no longer aggressively throw objects at people when malfunctioning. Instead they shudder and drop them on the ground creating mess. Makes the event a little bit less annoying.
- Space heater default temperature changed from 50C to 20C, which is the station's standard. Given heaters can also cool the air it's a quality of life improvement so you don't have to re-adjust the settings when using a new heater to cool an area back to normal temperature.
- Changelog entry

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2016-08-14-CellrateChange

atlantiscze commented:
This PR has been finished. Assuming the travis build passes it should be ready for review and merge.

A note about the power cell changes. While the subtype capacities were changed, some things do not use the subtypes and instead create power cells of custom capacity (which have to be manually adjusted). While i have done all i could to locate such items, it is possible i missed some. This mostly applies to various exotic items that are rarely used, and are only manually spawned/created later in the round. As i can not manually check every single item in the game, i assume this will have to do.

Even if i missed something, the only problem will be that the item will most likely have very large power cell in comparison to normal cells, which may be a bit weird, but not a game breaking issue.

PsiOmegaDelta commented on code/game/machinery/vending.dm:
```diff
@@ -573,21 +573,17 @@
-	src.visible_message("<span class='warning'>\The [src] launches \a [throw_item] at \the [target]!</span>")
+
+	dispensed_item.forceMove(get_turf(src))
+	visible_message("<span class='warning'>\The [src] shudders and \a [dispensed_item] falls out!</span>")
```
These unmentioned changes are relevant to cell capacities how?

**Edit:** Appears to be mentioned in the changelog, still unmentioned in the PR itself.

Raptor1628 replied:
This is NOT a tiny change, and I'd prefer it reverted. Basically invalidates the event, was wondering where this snuck in.

PsiOmegaDelta commented on code/modules/power/cell.dm:
```diff
@@ -337,6 +123,137 @@ var/cell_uid = 1		// Unique ID of this power cell. Used to reduce bunch of uglie
+/obj/item/weapon/cell/device/variable/New(newloc, charge_amount)
+	..(newloc)
+	maxcharge = charge_amount
+	charge = maxcharge
```
`charge = isnull(charge) ? maxcharge : charge` would allow providing a custom charge level as a type var assignment instead of using New() overrides.

atlantiscze replied:
This particular piece of code wasn't actually modified, i believe. It was only moved from one part of the file to other one. I can add the change if you want, but i can see no difference here, as the intention is that these cells come fully pre-charged and with custom capacity.

PsiOmegaDelta replied:
Very well.

As for why, as stated it would make it possible to write the empty definitions as

```
/obj/item/weapon/cell/empty
    charge = 0
```

```
/obj/item/weapon/cell/half_empty
    charge = 500
```

and it would no longer be necessary to override New() which could cause issues down the line if the base New() was changed somehow.

Techhead0 replied:
`charge = isnull(charge) ? maxcharge : 0` is wrong, isn't it?
It should be `charge = isnull(charge) ? maxcharge : charge`.

atlantiscze commented on code/game/machinery/vending.dm:
```diff
@@ -573,21 +573,17 @@
-	src.visible_message("<span class='warning'>\The [src] launches \a [throw_item] at \the [target]!</span>")
+
+	dispensed_item.forceMove(get_turf(src))
+	visible_message("<span class='warning'>\The [src] shudders and \a [dispensed_item] falls out!</span>")
```
This particular change was made according to feedback i've observed from players, that the vendor malfunction event is more often an outright annoying thing rather than something remotely useful/entertaining. This change still allows vendors to dispense items for free when malfunctioning, but without the [Common] Urist McChef says "HELP TO KITCHEN DINNERWARE IS KILLING ME!!" part.

Given it's a tiny change, i didn't really think this would need a completely separate PR. It is mentioned in the changelog, though i admit i must have forgotten to write it into the comment, as it was more or less a last minute change.

PsiOmegaDelta commented on code/game/objects/items/weapons/stunbaton.dm:
```diff
@@ -16,7 +16,7 @@
 	var/status = 0		//whether the thing is on or not
 	var/obj/item/weapon/cell/bcell = null
-	var/hitcost = 1000	//oh god why do power cells carry so much charge? We probably need to make a distinction between "industrial" sized power cells for APCs and power cells for everything else.
+	var/hitcost = 100	//oh god why do power cells carry so much charge? We probably need to make a distinction between "industrial" sized power cells for APCs and power cells for everything else.
```
Would've been practical to apply a constant to the various charge costs while you were at changing them, but I guess it's a bit much to redo.

PsiOmegaDelta commented:
> While the subtype capacities were changed, some things do not use the subtypes and instead create power cells of custom capacity (which have to be manually adjusted). While i have done all i could to locate such items, it is possible i missed some. 

And this is why I tell people to define and use appropriate subtypes instead of manually tweaking instances.

PsiOmegaDelta commented:
👍
